omcproxy: Rework omcproxy.init to make it work again

howl opened the pull request:
Actual omcproxy,init file doesn't bring up omcproxy at boot. This version is based in igmpproxy.init but adapted to omcproxy config file structure.

Changes from the old init are:
Set up network triggers in service_triggers.
Use ubus calls to get network device from interface name.
Use ubus calls to get interface status.
Specify proper family and proto options in firewall rules.

This fixes https://bugs.openwrt.org/index.php?do=details&task_id=1972 where omcproxy will not start up at boot because triggers aren't set properly in that situation and the network interface is not up when omcproxy init file starts. This init version set ups triggers even when no interface is up for the ones defined in omcproxy's config so when interface is ready it brings up.

 Signed-off-by: David Santamaría Rogado <howl.nsp@gmail.com>

dedeckeh commented:
The PR has formal issues, no SoB  and no clear git commit description what is changed in the patch. Please have a look at https://openwrt.org/submitting-patches

howl commented:
My bad, that doesn't edit the commit message, there is a way to edit a commit message?

bobafetthotmail commented:
The github web interface can't do it, you need a local git folder, see the submitting patches article for details on setting it up.

Then do a 
**git commit --amend**
in your local git, it will open a text editor to change the commit message and then 
**git push -f**
to update the github PR

howl committed:
omcproxy: Rework omcproxy.init to make it work again

Actual omcproxy,init file doesn't bring up omcproxy at boot. This version
is based in igmpproxy.init but adapted to omcproxy config file structure.

Changes from the previous init file are:
-Set up network triggers in service_triggers.
-Use ubus calls to get network device from interface name.
-Use ubus calls to get interface status.
-Specify proper family and proto options in firewall rules.

This fixes https://bugs.openwrt.org/index.php?do=details&task_id=1972
where omcproxy will not start up at boot because triggers aren't set
properly in that situation and the network interface is not up when
omcproxy init file starts. This init version set ups triggers even when
no interface is up for the ones defined in omcproxy's config so when
interface is ready it brings up.

Signed-off-by: David Santamaría Rogado <howl.nsp@gmail.com>

howl commented:
@bobafetthotmail , I have practically no idea to use git in command line but I could do it with your help. I only had one problem by editing another commit but could guess that was a branch selection problem.

@dedeckeh had edit now, also with more changes like the number of characters per line and name_of_the_package: in the subject. If there is something more I have done wrong please don't hesitate to point me.I want to add something more to the init file like config file triggers but as is now works similar that the original one but not failing at boot.

dedeckeh commented on package/network/services/omcproxy/files/omcproxy.init:
```diff
@@ -1,62 +1,88 @@
+
+	[ -n "$uplink" ] || return
+
+	local status="$(ubus -S call "network.interface.$uplink" status)"
```
network.sh (/lib/functions/network.sh) has wrapper functions which obsoletes the need to do ubus network calls in scripts

howl replied:
Done.

dedeckeh commented on package/network/services/omcproxy/files/omcproxy.init:
```diff
@@ -1,62 +1,88 @@
+	[ -n "$status" ] || return
+
+	json_load "$status"
+	json_get_var device l3_device
```
Retrieve the l3_device by using network_get_device

dedeckeh commented on package/network/services/omcproxy/files/omcproxy.init:
```diff
@@ -1,62 +1,88 @@
+
+	json_load "$status"
+	json_get_var device l3_device
+	json_get_var up up
```
You can check if an interface is up by using the network_is_up

dedeckeh commented on package/network/services/omcproxy/files/omcproxy.init:
```diff
@@ -1,62 +1,88 @@
+	proxy="$device"
+
+	for downlink in $downlinks; do
+		local status="$(ubus -S call "network.interface.$downlink" status)"
```
As mentioned above there's no need to use ubus network calls directly if you use the network.sh wrapper functions

dedeckeh commented on package/network/services/omcproxy/files/omcproxy.init:
```diff
@@ -1,62 +1,88 @@
 
-	proxy=""
+		json_load "$status"
+		json_get_var device l3_device
```
Use network_get_device() to retrieve the layer3 device

dedeckeh commented on package/network/services/omcproxy/files/omcproxy.init:
```diff
@@ -1,62 +1,88 @@
-	proxy=""
+		json_load "$status"
+		json_get_var device l3_device
+		json_get_var up up
```
Use network_is_up to check if an interface is up

dedeckeh commented on package/network/services/omcproxy/files/omcproxy.init:
```diff
@@ -1,62 +1,88 @@
 
-	for network in $uplink $downlink; do
-		procd_add_interface_trigger "interface.*" $network /etc/init.d/omcproxy restart
+	procd_add_interface_trigger "interface.*" $uplink /etc/init.d/omcproxy reload
```
I prefer to keep the procd_add_interface_trigger for the uplink in the for loop as before

dedeckeh commented on package/network/services/omcproxy/files/omcproxy.init:
```diff
@@ -104,14 +130,13 @@ omcproxy_add_firewall() {
 }
 
 start_service() {
-	include /lib/functions
```
This should not be removed if you make use of network.sh wrapper functions

howl committed:
omcproxy: Rework omcproxy.init to make it work again v2
Use network functions instead ubus calls to get device.
Omit getting status as network_get_device can be check if is null
for the same purpose.
Merge uplink and downlinks in just one for loop.

Signed-off-by: David Santamaría Rogado <howl.nsp@gmail.com>

howl commented:
@dedeckeh I did it with network functions the first time but changed it because testing with ash could see that network status was cached till calling a function to refresh the statuses returned by network functions. I should have asked before to not work two times.

I have pushed another commit, now network functions are used. There is no need to check status as I could see that getting device name results null if interface doesn't exist or is not up.

Also uplink merged in downlinks for as was in the original one.

I will see to change also igmpproxy as it use ubus calls too, but I need to get more git skills.

As before, anything that need to be changed just ask for it.

I have tested the new version and still works as expected.

dedeckeh commented:
@howl The PR has now 2 commits which makes the PR impossible to merge.
The first commit has the original changes while the second commit fixes my remarks which should have been in the original commit.
You should really try to stay away from the github user interface and rather fork the project to a public repo and clone the repo to a computer.
Instructions and guidance is provided in https://openwrt.org/submitting-patches#working_with_github

howl commented:
@dedeckeh tested your version and it works right, setting repeated interfaces and using PROCD_DEBUG=1 shows that your code works right avoiding repeated triggers. But, I could see that firewall rules anyways get repeated, but at first sight I don't know a way to do the same as you did for the triggers with the firewall rules without setting up another more global variable.

It could be commit as is now because it's better than the actual init script and it could be improved over the time.

The only little thing I think could be changed is in omcproxy_add_proxy(), moving "config_get downlinks $1 downlink" just before the downlinks for and "config_get scope $1 scope" just before "[ -n "$scope" ] && proxy="$proxy,scope=$scope"" to avoid getting configuration variables from proxies where no upstream is defined.

Also in omcproxy_add_network_triggers() a "[ -n "$uplink" ] || return" should be added just after getting upstream config variable, as triggers for no valid proxies are added.

If you think that are interesting changes I could make a patch.

dedeckeh commented:
I will close this PR as a reworked commit has been pushed (https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=df8f8bad082ea5280c19c9091c099b28b7e2fc67). I had to undo the change in omcproxy_add_proxy to set proxy=$device as it would break the check [ -n "$proxy" ] || return 0 if no downlink devices are found
